roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala

```diff
@@ -74,6 +78,35 @@ object VirtualMachine {
         (Left(error), errorState)
     }
 
+  def handleException(v: Ob, op: Op, tag: Location): Unit =
+    v.sysval match {
+      case SyscodeUpcall =>
+        op match {
+          case _: OpApplyCmd | _: OpApplyPrimArg | _: OpApplyPrimReg |
+              _: OpApplyPrimTag =>
+            handleApplyPrimUpcall(op, tag)
+          case _ =>
+            handleXmitUpcall(op, tag)
+        }
+
+      case SyscodeSuspend =>
+        op match {
+          case _: OpApplyCmd | _: OpApplyPrimArg | _: OpApplyPrimReg |
```

Comment:
Maybe it is necessary to extract type checking to another function to get rid of repetitions?

Reply:
Good call. But I think since the repetition just happens once it is fine to do it this way.